### DISCUSSION TITLE
How Do You Connect to an Existing Azure SQL Database?

### DISCUSSION THREAD

**Discussion by myokeeh** [Azure/azure-mobile-apps#267] [Q&A] [answered]
@adrianhall, I'm following this guide: https://adrianhall.github.io/azure/2021/11/08/azure-mobile-apps-intro/

Do you have guidance on connecting it to an existing Azure SQL Database? The guide makes it easy to get started by converting to `InMemoryTableData`, but I'm looking to test with the existing database I have from the old service. The new service apps runs fine with the `InMemoryTableData`.

I did attempt it, but the service fails to run with connection strings filled in for Azure SQL Database.

Detailed Error Information:
Module	   AspNetCoreModuleV2
Notification	   ExecuteRequestHandler
Handler	   aspNetCore
Error Code	   0x8007023e

**Comment by adrianhall** [comment, marked as answer]
So, there is a bug I'm working on now, which will prevent replacements and some queries from working.   However, it sounds like it's just not working for you, and you should be able to do the basics.

1) Create a new ASP.NET Core Datasync project using the template.
2) Set up the DbContext so that it reads the connection string from your original service.
3) Set up the Controller as before, but pass in the DbContext (e.g. ApplicationDbContext or whatever you use)
4) Use Repository = new EntityTableRepository<Model>(ApplicationDbContext)

There is a difference between the ITableData in the original code and the new ITableData.  This is to support more database types than before (e.g. Cosmos, MySQL, etc.)  Specifically, UpdatedAt was managed by the database before, and it's managed by the service now.  If you do this, you should probably create a copy of the EntityTableRepository.cs from the EFCore project, and remove the setting of UpdatedAt from it.  I'm still arguing with myself about how to do this in a backwards compatible way.

Finally, there are two bugs currently:

1) (which I discovered as part of #261), when using a trigger that does AFTER UPDATE, you need to copy the values of the metadata that are automatically set back to the entity that is the source of a replacement.  I've checked in the code, but not produced the NuGet yet.

2) (which is logged as #259), when you have certain queries that are more complex, EF Core throws an InvalidOperationException regarding client-side execution.  We don't trap this, when we should adjust the query accordingly.

Next release will fix both of these.

If it's none of this, I need more information about the bug.

**Reply by myokeeh** [reply]
Thanks for the detailed response. I'll give it a try.

FYI, I had to create the App Service manually via the portal since I was getting this error via Azure CLI:

>az webapp up --name [redacted_name] --location westus --sku FREE --runtime "DOTNET|6.0"

>'6.0' is not recognized as an internal or external command, operable program or batch file.

**Reply by adrianhall** [reply]
Likely, you need to update your Azure CLI.  I tested using the latest version.

**Reply by myokeeh** [reply]
I was able to get this up and running with some help from this page for 4: https://docs.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/server/dotnet-core

`az upgrade` indicates I am on 2.30.0 already. Weird. I didn't bother trying again since I got it working on the one I spun up yesterday.

On to more testing... probably MSAL auth

**Reply by adrianhall** [reply]
Awesome - let me know any problems you run into!

For MSAL auth, if you are using the older client, create a DelegatingHandler to add the authentication token to the request.  You may like to use the GenericAuthenticationProvider class (which is a delegating handler that asks for a token then adds it to the request).  When you new up the client, you can use:

``` csharp
var authProvider = new GenericAuthenticationProvider(() => GetTokenAsync());
var client = new AzureMobileClient(endpoint, authProvider);
```

Then somewhere else:

``` csharp
public async Task<AuthenticationToken> GetTokenAsync()
{
// Do whatever you need to do, and then create an AuthenticationToken with the right info
}
```

Both AuthenticationToken and GenericAuthenticationProvider are from https://github.com/Azure/azure-mobile-apps/tree/main/sdk/dotnet/src/Microsoft.Datasync.Client/Authentication

Hope that helps!  If it works, post the GetTokenAsync for all to see!

**Reply by myokeeh** [reply]
Not sure if should start a new discussion. My apps make use of the SQLite local sync and MSAL auth. I'm not necessarily tied to the old client. In fact, I'd rather experiment with newer stuff. I wasn't quite sure the state of things. What do you recommend? Is there a template/sample app for a WindowsAppSDK client? Is "azure-mobile-apps/samples/TodoApp/TodoApp.WPF" a good one to reference?

**Reply by adrianhall** [reply]
The samples in this repo are all for the new client, which doesn't have SQLite support yet (it's next on the list).  My intent is to have a specific sample for each environment - the basics (WPF, UWP, Xamarin.Forms, Xamarin.iOS, Xamarin.Android) and common frameworks (Avalonia, Uno).  I'm also looking at what it will take to support Blazor with a local cache based on IndexedDb.

The current client does online stuff + authentication - offline stuff is not ready yet.

**Reply by myokeeh** [reply]
Based on the code for the old client you provided above, replace `AzureMobileClient` with `DatasyncClient` like so to implement authentication, right?

```cs
var authProvider = new GenericAuthenticationProvider(() => GetTokenAsync());
var client = new AzureMobileClient(endpoint, authProvider);
```

Now, off to figure out `GetTokenAsync()`. Hopefully, mostly copy paste from my old client.

**Reply by myokeeh** [reply]
I've republished the service to enable authentication and tried the following:

In the old client, this is what I use. Is there a way to accomplish the same thing with the new one? 

```cs
payload = new JObject
{
    { "access_token",  msalAuthenticationResult.AccessToken }
};

mobileServiceClient.CurrentUser = await mobileServiceClient.LoginAsync(MobileServiceAuthenticationProvider.WindowsAzureActiveDirectory, payload);
```
With the new client, I'm now able to successfully log-in with MSAL and it provides me with an `AuthenticationResult`. I'm returning the following from the `GetTokenAsync()`, but the app now crashes with an XAML-related exception that's not too helpful.
```cs
public async Task<AuthenticationToken> GetTokenAsync()
{
    //...MSAL log-in
    return new AuthenticationToken() { DisplayName = msalAuthenticationResult.Account.Username, UserId = msalAuthenticationResult.UniqueId, Token = msalAuthenticationResult.AccessToken };
}
```
Exception:
>The application called an interface that was marshalled for a different thread. (Exception from HRESULT: 0x8001010E (RPC_E_WRONG_THREAD))

**Reply by adrianhall** [reply]
The MSAL piece needs to be called from the UI thread, probably.  The line you are showing isn't thread-specific, but UI things need to be started from the UI thread.

**Reply by adrianhall** [reply]
This might help: https://docs.microsoft.com/en-us/xamarin/essentials/main-thread

**Reply by myokeeh** [reply]
Running this on Desktop, but point taken. I'm checking on thread.

I've reverted to `_client = new DatasyncClient("https://yada.azurewebsites.net")` (without the `authProvider` and none of the MSAL code running) and it gives the same error. The only change I suspect currently is the addition of authentication to the service. Client app might be getting the error because of the 401 from the service.

I'm debugging with REST client with a valid AccessToken, but I'm also getting a 401. Sending the following:

GET https://yada.azurewebsites.net/tables/someitem

ZUMO-API-VERSION 3.0.0
X-ZUMO-AUTH [accesstoken_redacted]
X-ZUMO-FEATURES QS,OL
X-ZUMO-INSTALLATION-ID [some GUID]

**Reply by adrianhall** [reply]
That, at least, gives you a starting point.  You can't just include an MSAL auth token in X-ZUMO-AUTH.  In general, you have to either round-trip the token via the App Service Authentication piece to get a zumo token, or you have to integrate Microsoft.Identity.Web and submit the token via "Authorization: Bearer" header (in which case, your ASP.NET Core code deals with it).

Start by dropping the datasync services out - just get a simple API working with MSAL.  Once you have that, you can add the table code back in.

**Reply by mjfusa** [reply]
@adrianhall Per your comment above, can you confirm if Comos DB is supported in the latest release? If not, when do you think this support will be available?  Thanks

**Reply by adrianhall** [reply]
Latest beta supports anything EFCore supports.  That includes CosmosDB.  That being said, I haven't tried it with Cosmos yet - might do that tomorrow.

**Comment by myokeeh** [comment]
@adrianhall, I'm encouraged that my attempt last night got my existing Azure SQL Database connected the new service app and a UWP test app working based on the latest docs. This includes authentication with an Azure AD tenant and offline syncing!
